### PR TITLE
Implement TsuMeta Committee Plug-in

### PR DESCRIPTION
**Pull request by DeathlyPlays**

### CODE REVIEW

**Comment by DeathlyPlays** [comment]
In my 2nd block I will be rigorously testing this off a fresh clone of WL

**Comment by DeathlyPlays** [comment]
@HoeenCoder completed and meets @DesoGit's criteria

**Comment by DeathlyPlays** [comment]
I'm gonna toss in a lil' something real quick though for profiles @DesoGit and I spoke of

File: wavelength-plugins/tsumeta.js [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+ * TsuMeta Committee Plug-in *
+ * Created by Insist         *
+ * Idea/Made for Desokoro    *
+ *****************************/
+
+"use strict";
+
+let proposals = {};
+
+const fs = require('fs');
```
Comment (HoeenCoder) [inline wavelength-plugins/tsumeta.js]: I want to start using lib/fs.js for all new plugins, and refactor the old ones later

Comment (DeathlyPlays) [reply wavelength-plugins/tsumeta.js]: Oh, okay.

**Comment by DeathlyPlays** [comment]
oh active proposals list gotcha :p

File: wavelength-plugins/tsumeta.js [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+			}
+		},
+
+		suspend: function (target, room, user) {
+			if (!this.canTalk()) return this.errorReply("You cannot do this while unable to talk.");
+			if (!target || target.length > 18) return this.errorReply(`You must specify a target, with a maximum of 18 characters.`);
+			let targetUser = toId(target);
+			if (!isTsuMetaCouncil(targetUser) || targetUser === "desokoro") return this.errorReply(`${target} is either not in TsuMeta, or they have already suspended.`);
+			// Only allow xcmr to suspend users if he is currently in the council
+			if (user.userid === "insist" && Db.councilmember.has("insist", 1)) {
```
[the remark this answers is not in the document]
Comment (DeathlyPlays) [reply wavelength-plugins/tsumeta.js]: Oh ahit mb thats from testing

File: wavelength-plugins/tsumeta.js [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+			let reply = `<b><u>Proposals (${Object.keys(proposals).length})</u></b><br />`;
+			for (let proposal in proposals) reply += `(<strong>${proposal}</strong>)<br />`;
+			this.sendReply(`|raw|<div class="infobox infobox-limited">${reply}</div>`);
+		},
+
+		suspend: function (target, room, user) {
+			if (!this.canTalk()) return this.errorReply("You cannot do this while unable to talk.");
+			if (!target || target.length > 18) return this.errorReply(`You must specify a target, with a maximum of 18 characters.`);
+			let targetUser = toId(target);
+			if (!isTsuMetaCouncil(targetUser) || targetUser === "desokoro") return this.errorReply(`${target} is either not in TsuMeta, or they have already suspended.`);
```
Comment (HoeenCoder) [inline wavelength-plugins/tsumeta.js]: they are already*

File: wavelength-plugins/tsumeta.js [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+			Db.councilmember.set(tsuMetaMember, 1);
+			this.sendReply(`|html|${WL.nameColor(tsuMetaMember, true)} has been successfully been added into the TsuMeta Council.`);
+			if (Users.get(tsuMetaMember)) Users(tsuMetaMember).popup(`|html|You have been added into the TsuMeta Council by ${WL.nameColor(user.name, true)}.`);
+		},
+
+		kick: "take",
+		remove: "take",
+		delete: "take",
+		take: function (target, room, user) {
+			if (user.userid !== "desokoro") return this.errorReply("You must be Desokoro to remove users into the council.");
```
Comment (HoeenCoder) [inline wavelength-plugins/tsumeta.js]: from the council*

File: wavelength-plugins/tsumeta.js [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+		meeting: "message",
+		alert: "message",
+		pm: "message",
+		message: function (target, room, user) {
+			if (user.userid !== "desokoro" && !this.can("bypassall")) return this.errorReply(`This command is reserved for Desokoro only.`);
+			if (!target) return this.parse("/tsumetahelp");
+			let councilMembers = Db.councilmember.keys();
+			for (let u in councilMembers) {
+				if (!Users(councilMembers[u]).connected) continue;
```
Comment (HoeenCoder) [inline wavelength-plugins/tsumeta.js]: make sure the user object exists before checking if the user is connected

File: wavelength-plugins/tsumeta.js [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+		unsuspend: function (target, room, user) {
+			if (!this.canTalk()) return this.errorReply("You cannot do this while unable to talk.");
+			if (!target || target.length > 18) return this.errorReply(`You must specify a target, with a maximum of 18 characters.`);
+			let targetUser = toId(target);
+			if (!isTsuMetaCouncil(targetUser) || targetUser === "desokoro") return this.errorReply(`${target} is either not in TsuMeta, or they have already suspended.`);
+			// Only allow xcmr to unsuspend users if he is currently in the council
+			if (user.userid === "xcmr" && Db.councilmember.has("xcmr", 1)) {
+				Db.councilmember.set(targetUser, 1);
+				this.sendReply(`You have successfully suspended ${target} from participating in TsuMeta Committee proposals.`);
```
Comment (HoeenCoder) [inline wavelength-plugins/tsumeta.js]: unsuspended

File: wavelength-plugins/tsumeta.js [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+			this.sendReply(`|raw|<div class="infobox infobox-limited">${reply}</div>`);
+		},
+
+		suspend: function (target, room, user) {
+			if (!this.canTalk()) return this.errorReply("You cannot do this while unable to talk.");
+			if (!target || target.length > 18) return this.errorReply(`You must specify a target, with a maximum of 18 characters.`);
+			let targetUser = toId(target);
+			if (!isTsuMetaCouncil(targetUser) || targetUser === "desokoro") return this.errorReply(`${target} is either not in TsuMeta, or they have already suspended.`);
+			// Only allow xcmr to suspend users if he is currently in the council
+			if (user.userid === "xcmr" && Db.councilmember.has("xcmr", 1)) {
```
Comment (HoeenCoder) [inline wavelength-plugins/tsumeta.js]: Db.file.has only takes one argument

File: wavelength-plugins/tsumeta.js [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+			Db.councilmember.set(targetUser, 2);
+		},
+
+		unsuspend: function (target, room, user) {
+			if (!this.canTalk()) return this.errorReply("You cannot do this while unable to talk.");
+			if (!target || target.length > 18) return this.errorReply(`You must specify a target, with a maximum of 18 characters.`);
+			let targetUser = toId(target);
+			if (!isTsuMetaCouncil(targetUser) || targetUser === "desokoro") return this.errorReply(`${target} is either not in TsuMeta, or they have already suspended.`);
+			// Only allow xcmr to unsuspend users if he is currently in the council
+			if (user.userid === "xcmr" && Db.councilmember.has("xcmr", 1)) {
```
Comment (HoeenCoder) [inline wavelength-plugins/tsumeta.js]: Db.file.has only takes one argument

File: wavelength-plugins/tsumeta.js [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+	tsumeta: {
+		give: function (target, room, user) {
+			if (user.userid !== "desokoro") return this.errorReply("You must be Desokoro to add users into the council.");
+			if (!target) return this.parse("/tsumetahelp");
+			let tsuMetaMember = toId(target);
+			if (tsuMetaMember.length > 18) return this.errorReply("Usernames cannot exceed 18 characters.");
+			if (isTsuMetaCouncil(tsuMetaMember)) return this.errorReply(`${tsuMetaMember} is already in the TsuMeta Council.`);
+			Db.councilmember.set(tsuMetaMember, 1);
+			this.sendReply(`|html|${WL.nameColor(tsuMetaMember, true)} has been successfully been added into the TsuMeta Council.`);
+			if (Users.get(tsuMetaMember)) Users(tsuMetaMember).popup(`|html|You have been added into the TsuMeta Council by ${WL.nameColor(user.name, true)}.`);
```
Comment (HoeenCoder) [inline wavelength-plugins/tsumeta.js]: just `Users` for consitancy

File: wavelength-plugins/tsumeta.js [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+		delete: "take",
+		take: function (target, room, user) {
+			if (user.userid !== "desokoro") return this.errorReply("You must be Desokoro to remove users into the council.");
+			if (!target) return this.parse(`/tsumetahelp`);
+			let tsuMetaMember = toId(target);
+			if (tsuMetaMember.length > 18) return this.errorReply("Usernames cannot exceed 18 characters.");
+			if (!isTsuMetaCouncil(tsuMetaMember)) return this.errorReply(`${tsuMetaMember} isn't a TsuMeta council member.`);
+			Db.councilmember.remove(tsuMetaMember);
+			this.sendReply(`|html|${WL.nameColor(tsuMetaMember, true)} has been removed from the TsuMeta council.`);
+			if (Users.get(tsuMetaMember)) Users(tsuMetaMember).popup(`|html|You have been removed from the TsuMeta Council by ${WL.nameColor(user.name, true)}.`);
```
Comment (HoeenCoder) [inline wavelength-plugins/tsumeta.js]: just `Users`

File: wavelength-plugins/tsumeta.js (hunk 5 above)
Comment (DeathlyPlays) [reply wavelength-plugins/tsumeta.js]: how?

Comment (HoeenCoder) [reply wavelength-plugins/tsumeta.js]: `if (!Users(councilMembers[u]) || !Users(councilMembers[u]).connected) continue;`

**Pull request merged**